Feature/530/migrate to plugin framework

dimuon opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

## Description
<!--- Describe your changes in detail. -->

Migration to [TF Plugin Framework](https://developer.hashicorp.com/terraform/plugin/framework).
New schema for `ec_deployment`.

*Limitations*
* State upgrade is not currently implemented for `ec_deployment`. The recommended way to port existing resources is importing.
* The migration is based on the provider version 0.4.1. We plan to bring changes from 0.5.0 in future releases.
* Few acceptance tests fail. Please see the testing section below.

## Related Issues
<!--- This project only accepts pull requests related to open issues. -->
<!--- If suggesting a new feature or change, please discuss it in an -->
<!--- issue first.  If fixing a bug, there should be an issue describing -->
<!--- it with steps to reproduce.  Please link to the any related issues -->
<!--- here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

* unit tests (majority of the relevant unit tests were migrated)
* acceptance tests
* manual testing

There are still few acceptance tests that fail:
- 4 new acceptance tests (their names mention version 0.4.1) that verify migration of other data sources and resources from 0.4.1 to 0.6.0. To fix them we need implement a state upgrade for `ec_deployment`.
- `TestAccDeploymentExtension_pluginDownload` fails on 0.4.1 with the same error ("Custom plugin is not allowed at your subscription level.")
- `TestAccDeploymentExtension_bundleFile` fails quite often on both 0.4.1 and the branch with the "failed to upload file" error.
- `TestAccDeployment_withExtension` fails quite often with the same error as `TestAccDeploymentExtension_bundleFile`.

<!--- Please describe in detail how you tested your changes.  Include -->
<!--- details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of Changes
<!--- What types of changes does your code introduce? Put an `x` in -->
<!--- all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (improves code quality but has no user-facing effect)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation

## Readiness Checklist
<!--- Go over all the following points, and put an `x` in all the boxes -->
<!--- that apply.  If you're unsure about any of these, don't hesitate -->
<!--- to ask. We're here to help! -->
- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [ ] All new and existing tests passed

tobio commented:
> The recommended way to port existing resources to the new provider version is [state import](https://developer.hashicorp.com/terraform/cli/import#state-only).

It's not possible to import the elastic password, this feels like a pretty big foot gun that we're handing users.

tobio commented:
<img width="217" alt="image" src="https://user-images.githubusercontent.com/444668/206162164-b6c2495f-372a-4722-9b94-683b5de2b056.png">

To be honest, this is impossible to actually review. Can we instead break this up so we're only migrating a single resource/data source per PR?

dimuon commented:
> <img alt="image" width="217" src="https://user-images.githubusercontent.com/444668/206162164-b6c2495f-372a-4722-9b94-683b5de2b056.png">
> 
> To be honest, this is impossible to actually review. Can we instead break this up so we're only migrating a single resource/data source per PR?

Yes, it's quite difficult to review this. We did some reviews with @pascal-hofmann when we were working on the feature branch and basically we followed the way that you proposed. But, even in that case, it's not so easy - `ec_deployment` code base is quite big by its own, e.g. [this](https://github.com/dimuon/terraform-provider-ec/pull/2) was the PR for `ec_deployment` only. However, there are also a number of json files for acceptance and unit tests.

tobio commented on ec/internal/converters/convert_tags.go:
```diff
@@ -0,0 +1,96 @@
+
+// flattenTags takes in Deployment Metadata resource models and returns its
+// Tags in flattened form.
+func TagsToTypeMap(metadataItems []*models.MetadataItem) types.Map {
```
We're using `*Types`, `TF*` and `*TF` to refer to the Terraform types in function names. IMO `*Types` is nicest, I'm not fussed which we use but we should name these functions consistently.

dimuon commented on README.md:
````diff
@@ -114,3 +120,119 @@ $ export EC_API_KEY="<apikey value>"
+}
+```
+
+Please note that the configuration explicitly mentions `hot` tier and the tier has `autoscaling` and `config` attributes even despite the fact that they are empty. If they were omitted, TF (at least up to version 1.3.3) could complain `Error: Provider produced inconsistent result after apply`.
````
👍  The sentence is outdated. The new schema doesn't have `config` on the tier level anymore. Also, `autoscaling` is required now.

dimuon commented on ec/provider.go:
```diff
@@ -59,105 +70,300 @@ var (
-	}
-}
-
 func newSchema() map[string]*schema.Schema {
```
Yes, I'll do, thanks for catching this. However, this would also require a refactoring of the test.

tobio commented on ec/ecdatasource/deploymentdatasource/flatteners_integrations_server.go:
```diff
@@ -18,84 +18,109 @@
 // flattened form.
-func flattenIntegrationsServerResources(in []*models.IntegrationsServerResourceInfo) []interface{} {
-	var result = make([]interface{}, 0, len(in))
+func flattenIntegrationsServerResources(ctx context.Context, in []*models.IntegrationsServerResourceInfo, target interface{}) diag.Diagnostics {
```
Should return the model rather than assign to interface pointer.

dimuon replied:
@pascal-hofmann ?

tobio commented on ec/ecdatasource/deploymentdatasource/flatteners_kibana.go:
```diff
@@ -18,83 +18,109 @@
-	var result = make([]interface{}, 0, len(in))
-	for _, res := range in {
-		var m = make(map[string]interface{})
+func flattenKibanaResources(ctx context.Context, in []*models.KibanaResourceInfo, target interface{}) diag.Diagnostics {
```
Samesies

dimuon replied:
@pascal-hofmann ?

tobio commented on ec/ecdatasource/deploymentdatasource/flatteners_observability.go:
```diff
@@ -17,33 +17,54 @@
-func flattenObservability(settings *models.DeploymentSettings) []interface{} {
-	if settings == nil || settings.Observability == nil {
-		return nil
+func flattenObservability(ctx context.Context, settings *models.DeploymentSettings, target interface{}) diag.Diagnostics {
```
Samesies

dimuon replied:
@pascal-hofmann ?

tobio commented on ec/ecdatasource/deploymentdatasource/flatteners_traffic_filter.go:
```diff
@@ -18,19 +18,22 @@
 
 // flattenTrafficFiltering parses a deployment's traffic filtering settings.
-func flattenTrafficFiltering(settings *models.DeploymentSettings) []interface{} {
+func flattenTrafficFiltering(ctx context.Context, settings *models.DeploymentSettings, target interface{}) diag.Diagnostics {
```
samesies

dimuon replied:
@pascal-hofmann ?

tobio commented on ec/ecdatasource/deploymentdatasource/schema.go:
```diff
@@ -18,103 +18,79 @@
+			},
+			"region": {
+				Type:        types.StringType,
+				Description: "Region where the deployment can be found.",
```
```suggestion
				Description: "Region where the deployment is hosted.",
```

tobio commented on ec/ecdatasource/deploymentdatasource/schema.go:
```diff
@@ -18,103 +18,79 @@
-				Computed: true,
+			"deployment_template_id": {
+				Type:        types.StringType,
+				Description: "ID of the deployment template used to create the deployment.",
```
```suggestion
				Description: "ID of the deployment template this deployment is based off.",
```

Deployment templates can be changed so we shouldn't talk about creation here.

tobio commented on ec/ecdatasource/deploymentdatasource/schema_apm.go:
```diff
@@ -18,72 +18,116 @@
-				Type:     schema.TypeString,
-				Computed: true,
+				Type:        types.StringType,
+				Description: "The user-specified ID of the Elasticsearch cluster to which this resource kind will link.",
```
Can we update any references of `resource kind` to `APM resource`

dimuon replied:
@pascal-hofmann ? I fixed this description, but there are still other similar places in the file.

tobio commented on ec/ecdatasource/deploymentdatasource/schema_apm.go:
```diff
@@ -18,72 +18,116 @@
+			},
+			"size": {
+				Type:        types.StringType,
+				Description: "Amount of resource per topology element in the \"g\" notation.",
```
What's the "g" notation. 

```suggestion
				Description: "Amount of size_resource in Gigabytes. For example \"4g\".",
```

tobio commented on ec/ecdatasource/deploymentdatasource/flatteners_elasticsearch.go:
```diff
@@ -18,147 +18,172 @@
 
-				for k, v := range util.FlattenClusterEndpoint(res.Info.Metadata) {
-					m[k] = v
+				endpoints := util.FlattenClusterEndpoint(res.Info.Metadata)
```
Actually that already exists in `converters.ExtractEndpointsTF`, we should remove `FlattenClusterEndpoint` and use the converter.

tobio commented on ec/ecresource/deploymentresource/apm/v2/apm.go:
```diff
@@ -0,0 +1,199 @@
+	return &apm, nil
+}
+
+func (apm ApmTF) Payload(ctx context.Context, payload models.ApmPayload) (*models.ApmPayload, diag.Diagnostics) {
```
Should be called something like `EnrichTemplate`

dimuon replied:
My take was that the func generates payload for Apm based on initial payload from a template. It doesn't matter how the func does it.

tobio commented on ec/ecresource/deploymentresource/apm/v2/apm.go:
```diff
@@ -0,0 +1,199 @@
+	return &payload, diags
+}
+
+func ApmPayload(ctx context.Context, apmObj types.Object, template *models.DeploymentTemplateInfoV2) (*models.ApmPayload, diag.Diagnostics) {
```
Similarly, this is `EnrichTemplate` or something

dimuon replied:
From a caller perspective the func generates payload while `EnrichTemplate` describes more implementation details of the func.

tobio commented on ec/ecresource/deploymentresource/delete.go:
```diff
@@ -20,82 +20,84 @@ package deploymentresource
-func shouldRetryShutdown(err error, retries, maxRetries int) bool {
-	const timeout = "Timeout exceeded"
-	needsRetry := retries < maxRetries
+func removeRule(ruleID, deploymentID string, client *api.API) error {
```
IMO this doesn't make sense here. It's not called when deleting the deployment resource, it should live with the update code.

tobio commented on ec/ecresource/deploymentresource/create.go:
```diff
@@ -21,62 +21,76 @@ import (
-	"github.com/elastic/cloud-sdk-go/pkg/multierror"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	deploymentv "github.com/elastic/terraform-provider-ec/ec/ecresource/deploymentresource/deployment/v2"
```
Looks like we can drop the import alias here.

tobio commented on ec/ecresource/deploymentresource/create.go:
```diff
@@ -21,62 +21,76 @@ import (
 	}
 
-	d.SetId(*res.ID)
+	tflog.Trace(ctx, "created a resource")
```
```suggestion
	tflog.Trace(ctx, "created deployment resource")
```

tobio commented on ec/ecresource/deploymentresource/utils/definitions.go:
```diff
@@ -15,24 +15,9 @@
-import (
-	"github.com/elastic/cloud-sdk-go/pkg/models"
+const (
+	// minimumEnterpriseSearchSize   = 2048
```
Remove the comment

tobio commented on ec/ecresource/deploymentresource/utils/get_first.go:
```diff
@@ -15,27 +15,30 @@
-	return &schema.Resource{
-		Description: "Elastic Cloud deployment Elasticsearch keystore",
-		Schema:      newSchema(),
+func GetFirst(ctx context.Context, list types.List, target any) diag.Diagnostics {
```
This is unused, can we delete it.

tobio commented on ec/ecresource/extensionresource/read.go:
```diff
@@ -60,40 +81,40 @@ func extensionNotFound(err error) bool {
+	if model.ExtensionType != nil {
+		state.ExtensionType = types.String{Value: *model.ExtensionType}
+	} else {
+		state.ExtensionType = types.String{Null: true}
 	}
 
-	if err := d.Set("description", model.Description); err != nil {
-		return err
-	}
+	state.Description = types.String{Value: model.Description}
```
Why do we treat these attributes differently? Why are we explicitly setting these values to null in this resource?

dimuon replied:
@pascal-hofmann , can you please help with an answer?

pascal-hofmann replied:
Pointer vs. non-pointer.

tobio commented on ec/ecresource/extensionresource/schema.go:
```diff
@@ -18,64 +18,151 @@
+				Description: "download url",
+				Computed:    true,
+				Optional:    true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{
+					planmodifier.DefaultValue(types.String{Value: ""}),
+				},
```
Why does this make sense as a default?

dimuon replied:
@pascal-hofmann , can you please help with an answer?

pascal-hofmann replied:
This is the default from the upstream API. Plan modifier prevents difference between planned state and result after apply.

tobio commented on ec/ecresource/extensionresource/schema.go:
```diff
@@ -18,64 +18,151 @@
+				Type:        types.StringType,
+				Description: "Description for extension",
+				Computed:    true,
+				Optional:    true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{
+					planmodifier.DefaultValue(types.String{Value: ""}),
+				}},
```
Similarly, why are we adding a default here?

dimuon replied:
@pascal-hofmann , can you please help with an answer?

tobio commented on ec/ecresource/trafficfilterresource/flatteners.go:
```diff
@@ -18,66 +18,66 @@
 
-func flattenRules(rules []*models.TrafficFilterRule) *schema.Set {
-	result := schema.NewSet(trafficFilterRuleHash, []interface{}{})
+func flattenRules(ctx context.Context, rules []*models.TrafficFilterRule, target interface{}) diag.Diagnostics {
```
Would be nicer to return a typed model rather than assigning to an interface pointer.

dimuon replied:
@pascal-hofmann ?

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v1/schema.go:
```diff
@@ -0,0 +1,544 @@
+				Type:        types.StringType,
+				Description: "The Elasticsearch resource unique identifier",
+				Computed:    true,
+				// PlanModifiers: tfsdk.AttributePlanModifiers{
```
Can we remove the comments here please.

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v1/schema.go:
```diff
@@ -0,0 +1,544 @@
+				Computed:    true,
+				PlanModifiers: tfsdk.AttributePlanModifiers{
+					resource.UseStateForUnknown(),
+					resource.RequiresReplaceIf(func(ctx context.Context, state, config attr.Value, path path.Path) (bool, diag.Diagnostics) {
```
This should be user specified, why do we need `RequiresReplace*` on this attribute?

dimuon replied:
I don't think plan modifiers in v1 are useful at all - I'm going to delete this.

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v1/schema.go:
```diff
@@ -0,0 +1,544 @@
+		Attributes: tfsdk.ListNestedAttributes(map[string]tfsdk.Attribute{
+			"autoscale": {
+				Type:        types.StringType,
+				Description: `Enable or disable autoscaling. Defaults to the setting coming from the deployment template. Accepted values are "true" or "false".`,
```
Should this been a boolean attribute rather than a string?

dimuon replied:
This is how it was before. But once we're changing the schema anyway, maybe it makes sense to convert it to boolean one.

dimuon replied:
Done

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v1/schema.go:
```diff
@@ -0,0 +1,544 @@
+			// DiffSuppressFunc: suppressMissingOptionalConfigurationBlock,
+			"docker_image": {
+				Type:        types.StringType,
+				Description: "Optionally override the docker image the Elasticsearch nodes will use. Note that this field will only work for internal users only.",
```
```suggestion
				Description: "Optionally override the docker image the Elasticsearch nodes will use. This option will not work in ESS customers and should only be changed if you know what you're doing.",
```

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v1/schema.go:
```diff
@@ -0,0 +1,544 @@
+		Optional:    true,
+		Validators:  []tfsdk.AttributeValidator{listvalidator.SizeAtMost(1)},
+		Attributes: tfsdk.ListNestedAttributes(map[string]tfsdk.Attribute{
+			// TODO
```
Does this need to be done?

dimuon replied:
I believe no.

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/elasticsearch.go:
```diff
@@ -0,0 +1,413 @@
+		return nil, diags
+	}
+
+	if es == nil {
+		return nil, nil
+	}
+
+	if es == nil {
+		var diags diag.Diagnostics
+		diags.AddError("Elasticsearch payload error", "cannot find elasticsearch data")
+		return nil, diags
+	}
```
Only need one of these conditions.

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/elasticsearch.go:
```diff
@@ -0,0 +1,413 @@
+	return nil, nil
+}
+
+func ReadElasticsearch(in *models.ElasticsearchResourceInfo, remotes *models.RemoteResources) (*Elasticsearch, error) {
```
This doesn't need to be exported. 

```suggestion
func readElasticsearch(in *models.ElasticsearchResourceInfo, remotes *models.RemoteResources) (*Elasticsearch, error) {
```

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/elasticsearch.go:
```diff
@@ -0,0 +1,413 @@
+	return &es, nil
+}
+
+func (es *ElasticsearchTF) Payload(ctx context.Context, res *models.ElasticsearchPayload, skipTopologies bool) (*models.ElasticsearchPayload, diag.Diagnostics) {
```
Doesn't need to be exported

```suggestion
func (es *ElasticsearchTF) payload(ctx context.Context, res *models.ElasticsearchPayload, skipTopologies bool) (*models.ElasticsearchPayload, diag.Diagnostics) {
```

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/elasticsearch.go:
```diff
@@ -0,0 +1,413 @@
+	}
+}
+
+func UpdateNodeRolesOnDedicatedTiers(topologies []*models.ElasticsearchClusterTopologyElement) {
```
Doesn't need to be exported 

```suggestion
func updateNodeRolesOnDedicatedTiers(topologies []*models.ElasticsearchClusterTopologyElement) {
```

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/elasticsearch_config.go:
```diff
@@ -0,0 +1,120 @@
+	return c == nil || reflect.ValueOf(*c).IsZero()
+}
+
+func ReadElasticsearchConfig(in *models.ElasticsearchConfiguration) (*ElasticsearchConfig, error) {
```
Doesn't look like it needs to be exported 

```suggestion
func readElasticsearchConfig(in *models.ElasticsearchConfiguration) (*ElasticsearchConfig, error) {
```

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/elasticsearch_extension.go:
```diff
@@ -0,0 +1,136 @@
+
+type ElasticsearchExtensions v1.ElasticsearchExtensions
+
+func ReadElasticsearchExtensions(in *models.ElasticsearchConfiguration) (ElasticsearchExtensions, error) {
```
Doesn't look like it needs to be exported. 

```suggestion
func readElasticsearchExtensions(in *models.ElasticsearchConfiguration) (ElasticsearchExtensions, error) {
```

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/elasticsearch_config.go:
```diff
@@ -0,0 +1,120 @@
+	return &config, nil
+}
+
+func ElasticsearchConfigPayload(ctx context.Context, cfgObj attr.Value, model *models.ElasticsearchConfiguration) (*models.ElasticsearchConfiguration, diag.Diagnostics) {
```
Doesn't look like it needs to be exported 

```suggestion
func elasticsearchConfigPayload(ctx context.Context, cfgObj attr.Value, model *models.ElasticsearchConfiguration) (*models.ElasticsearchConfiguration, diag.Diagnostics) {
```

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/topology_plan_modifier.go:
```diff
@@ -0,0 +1,112 @@
+	tier string
+}
+
+func (m useTierState) Modify(ctx context.Context, req tfsdk.ModifyAttributePlanRequest, resp *tfsdk.ModifyAttributePlanResponse) {
```
Where are the tests for this?

dimuon replied:
I added tests.

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/node_types_plan_modifier.go:
```diff
@@ -0,0 +1,124 @@
+
+type nodeTypesDefault struct{}
+
+func (r nodeTypesDefault) Modify(ctx context.Context, req tfsdk.ModifyAttributePlanRequest, resp *tfsdk.ModifyAttributePlanResponse) {
```
Needs tests

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/node_roles_plan_modifier.go:
```diff
@@ -0,0 +1,106 @@
+
+type nodeRolesDefault struct{}
+
+func (r nodeRolesDefault) Modify(ctx context.Context, req tfsdk.ModifyAttributePlanRequest, resp *tfsdk.ModifyAttributePlanResponse) {
```
Tests

dimuon replied:
Done.

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/node_roles_plan_modifier.go:
```diff
@@ -0,0 +1,106 @@
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+// Use `self` as value of `observability`'s `deployment_id` attribute
```
This comment isn't correct

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/node_roles_plan_modifier.go:
```diff
@@ -0,0 +1,106 @@
+		return
+	}
+
+	if useNodeRoles && req.AttributeState.IsNull() {
```
This appears to be identical to `node_types_plan_modifier`. Should this condition be reversed for one of them? 

IMO we should likely extract this function so both modifiers can re-use it with presumably the opposite condition. 

Probably would be worthwhile having a comment on the modifier explaining the purpose here.

dimuon replied:
Done.

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/schema.go:
```diff
@@ -0,0 +1,479 @@
+		Required:    true,
+		Attributes: tfsdk.SingleNestedAttributes(map[string]tfsdk.Attribute{
+			"autoscale": {
+				Type:        types.StringType,
```
Can this not be a BoolType?

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/schema.go:
```diff
@@ -0,0 +1,479 @@
+		Optional:    true,
+		Attributes: tfsdk.SingleNestedAttributes(map[string]tfsdk.Attribute{
+			// TODO
+			// DiffSuppressFunc: suppressMissingOptionalConfigurationBlock,
```
Do the todo?

dimuon replied:
I'm going to remove all `DiffSuppressFunc` TODO comments. 
The Framework doesn't support such functions anymore. 
If `config` can be changed outside of TF, users can resort to `ingore_chages` meta attribute.

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/schema.go:
```diff
@@ -0,0 +1,479 @@
+func ElasticsearchTopologyAutoscalingSchema(tierName string) tfsdk.Attribute {
+	return tfsdk.Attribute{
+		Description: "Optional Elasticsearch autoscaling settings, such a maximum and minimum size and resources.",
+		// Optional:    true,
```
Can we also kill the commented code.

dimuon replied:
Done

tobio commented on ec/ecresource/deploymentresource/read.go:
```diff
@@ -20,25 +20,61 @@ package deploymentresource
+	}
+
+	var newState *deploymentv2.DeploymentTF
+	var err error
```
```suggestion
```

tobio commented on ec/ecresource/deploymentresource/read.go:
```diff
@@ -20,25 +20,61 @@ package deploymentresource
+	var err error
+
+	// use state for the plan (there is no plan and config during Read) - otherwise we can get unempty plan output
+	if newState, diags = r.read(ctx, curState.Id.Value, &curState, curState, nil); err != nil {
```
```suggestion
	if newState, diags = r.read(ctx, curState.Id.Value, &curState, curState, nil); diags.HasError() {
```

tobio commented on ec/ecresource/deploymentresource/read.go:
```diff
@@ -48,37 +84,86 @@ func readResource(_ context.Context, d *schema.ResourceData, meta interface{}) d
+
+	// ReadDeployment returns empty config struct if there is no config, so we have to nullify it if plan doesn't contain it
+	// we use state for plan in Read and there is no state during import so we need to check elasticsearchPlan against nil
+	if elasticsearchPlan != nil && elasticsearchPlan.Config.IsNull() && deployment.Elasticsearch != nil && deployment.Elasticsearch.Config != nil && deployment.Elasticsearch.Config.IsEmpty() {
```
Can we break this line up please.

tobio commented on ec/ecresource/deploymentresource/read.go:
```diff
@@ -92,3 +177,24 @@ func deploymentNotFound(err error) bool {
 	return apierror.IsRuntimeStatusCode(err, 403)
 }
+
+// Setting this variable here so that it is parsed at compile time in case
+// any errors are thrown, they are at compile time not when the user runs it.
```
```suggestion
// Setting this variable here so that it is parsed at compile time in case
// any errors are thrown, they are at compile time not when the user runs it.
```

This comment is inaccurate, this variable isn't parsed at compile time since it's not a compile time constant. It's executed when this package is first loaded. Granted, if this is invalid it will break in tests, but the comment should reflect that (setting this here so it breaks unit tests) rather than claim it's something that's checked in the compiler.

dimuon replied:
Thanks for the catch up. This is the leftover of the previous code that was moved as-is (though it looks like I renamed it :) ). I'll fix this.

tobio commented on ec/ecresource/deploymentresource/update.go:
```diff
@@ -19,73 +19,150 @@ package deploymentresource
-	for attr := range d.State().Attributes {
-		if strings.HasPrefix(attr, "traffic_filter") {
-			continue
+func handleTrafficFilterChange(ctx context.Context, client *api.API, plan, state v2.DeploymentTF) diag.Diagnostics {
```
Would be great to have some unit tests for these

dimuon replied:
I added tests.

tobio commented:
I'm concerned about the change to explicit tier attributes (hot, warm etc) and how that impacts ECE customers. Custom templates don't necessarily use the hot/warm/cold ids and nothing else in the platform requires that mapping. The previous approach (mapping based on a user supplied ID) provided the flexibility to support arbitrary custom templates. 

Are we being overly prescriptive here in such a way that impacts ECE customers? 

Is it not possible to have a set of topology elements keyed off the ID attribute instead? The attribute approach for topology elements feels like a step backwards.

tobio commented on ec/ecresource/deploymentresource/read.go:
```diff
@@ -48,37 +84,86 @@ func readResource(_ context.Context, d *schema.ResourceData, meta interface{}) d
+
+	var elasticsearchPlan *elasticsearchv2.ElasticsearchTF
+
+	if diags = tfsdk.ValueAs(ctx, plan.Elasticsearch, &elasticsearchPlan); diags.HasError() {
```
Why are we reading values from the state in this case. Shouldn't we be reading the API response here?

dimuon replied:
In this line we're reading the current plan. The API response is read in line 134.

dimuon replied:
@tobio you're correct in a sense that it could be state in some cases but otherwise my original answer still states true.

tobio replied:
All good, I'd confused myself.

tobio commented on ec/ecresource/deploymentresource/deployment/v2/deployment.go:
```diff
@@ -0,0 +1,566 @@
+		size, err := converters.ParseTopologySize(tier.Size, tier.SizeResource)
+
+		// we can ignore returning an error here - it's handled in readers
+		if err == nil && size != nil && size.Value != nil && *size.Value == 0 {
```
Are we reintroducing https://github.com/elastic/terraform-provider-ec/issues/471 here? Should we also be checking if the tier is autoscaleable?

dimuon replied:
I don't think so. The current approach is to specify all tiers (even if they are empty) if they have non-zero size in the used template or can be scaled up (that is the case when autoscaling is enabled). So in case of #471, the TF configuration can look like 

```
resource "ec_deployment" "test-471" {
  name                   = "test_471"
  region                 = var.ec_region
  version                = data.ec_stack.latest.version
  deployment_template_id = "aws-storage-optimized-v2"

  elasticsearch = {
    autoscale = "true"

    hot = {
      size       = "1g"
      zone_count = 3

      autoscaling = {
        max_size = "8g"
      }
    }

    ml = {
      zone_count = 1

      autoscaling = {
        min_size = "0g"
        max_size = "60g"
      }
    }

    warm = {
      autoscaling = {}
    }

    cold = {
      autoscaling = {}
    }

    frozen = {
      autoscaling = {}
    }
    
  }

  kibana = {
    size       = "1g"
    zone_count = 2
  }

  lifecycle {
    ignore_changes = [
      elasticsearch.hot.size,
      elasticsearch.warm.size,
      elasticsearch.cold.size,
      elasticsearch.frozen.size,
      elasticsearch.ml.size,
    ]
  }  

}

```

[README](https://github.com/dimuon/terraform-provider-ec/blob/feature/530/migrate-to-plugin-framework/README.md#moving-to-tf-framework-and-schema-change-for-ec_deployment-resource) and [docs](https://github.com/dimuon/terraform-provider-ec/blob/feature/530/migrate-to-plugin-framework/docs/resources/ec_deployment.md#tiered-deployment-with-autoscaling-enabled) mentions this approach. Still any suggestions are welcomed.

dimuon commented:
> I'm concerned about the change to explicit tier attributes (hot, warm etc) and how that impacts ECE customers. Custom templates don't necessarily use the hot/warm/cold ids and nothing else in the platform requires that mapping. The previous approach (mapping based on a user supplied ID) provided the flexibility to support arbitrary custom templates.
> 
> Are we being overly prescriptive here in such a way that impacts ECE customers?
> 
> Is it not possible to have a set of topology elements keyed off the ID attribute instead? The attribute approach for topology elements feels like a step backwards.

We discussed the approach with @Kushmaro. I think I tried using `set` but that didn't work out that time (as far I can recall we faced with non empty plan with computed elements of `set`), but it could be that I missed something.

My take is that generally the approach with an attribute per tier looks simpler with regard to topology manipulating and eliminate that bunch of defects/limitations that we had.

Kushmaro commented:
>I'm concerned about the change to explicit tier attributes (hot, warm etc) and how that impacts ECE customers. Custom templates don't necessarily use the hot/warm/cold ids and nothing else in the platform requires that mapping. The previous approach (mapping based on a user supplied ID) provided the flexibility to support arbitrary custom templates.

That's a valid concern, I wasn't aware that ECE diverged so far off when it comes to custom templates that can just have any "topology" kind. But I also recall no other way of doing this (as @dimuon pointed out) was solving for the issues we set out to solve, to begin with using topologies.

Also, how did Pascal not encounter this given he's testing against ECE?

What are our options here in such a case?

dimuon commented:
> > I'm concerned about the change to explicit tier attributes (hot, warm etc) and how that impacts ECE customers. Custom templates don't necessarily use the hot/warm/cold ids and nothing else in the platform requires that mapping. The previous approach (mapping based on a user supplied ID) provided the flexibility to support arbitrary custom templates.
> 
> That's a valid concern, I wasn't aware that ECE diverged so far off when it comes to custom templates that can just have any "topology" kind. But I also recall no other way of doing this (as @dimuon pointed out) was solving for the issues we set out to solve, to begin with using topologies.
> 
> Also, how did Pascal not encounter this given he's testing against ECE?
> 
> What are our options here in such a case?

I think initially Pascal proposed that approach with an attribute per tier.

pascal-hofmann commented:
> > Also, how did Pascal not encounter this given he's testing against ECE?
> > What are our options here in such a case?
> 
> I think initially Pascal proposed that approach with an attribute per tier.

I did not encounter this, as we just use the default tiers in our ECE instance.

dimuon commented:
> > I'm concerned about the change to explicit tier attributes (hot, warm etc) and how that impacts ECE customers. Custom templates don't necessarily use the hot/warm/cold ids and nothing else in the platform requires that mapping. The previous approach (mapping based on a user supplied ID) provided the flexibility to support arbitrary custom templates.
> 
> That's a valid concern, I wasn't aware that ECE diverged so far off when it comes to custom templates that can just have any "topology" kind. But I also recall no other way of doing this (as @dimuon pointed out) was solving for the issues we set out to solve, to begin with using topologies.
> 
> Also, how did Pascal not encounter this given he's testing against ECE?
> 
> What are our options here in such a case?

We can add `id` attribute to the tier schema that will help us with mapping custom ids to the provider's tier names. The provider supports 7 tiers now so as long as this is enough we shall be good.

tobio commented:
@pascal-hofmann are you able to sign the [CLA](https://www.elastic.co/contributor-agreement) so we're able to merge this when it's ready?

pascal-hofmann commented:
> @pascal-hofmann are you able to sign the [CLA](https://www.elastic.co/contributor-agreement) so we're able to merge this when it's ready?

Not yet. Our legal department is still checking the CLA.

dimuon commented on ec/provider.go:
```diff
@@ -59,105 +70,300 @@ var (
+				Type:          types.StringType,
+				Optional:      true,
+				Sensitive:     true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{planmodifier.DefaultFromEnv([]string{"EC_API_KEY"})},
```
I removed it too - plan modifiers don't apply to providers.

dimuon replied:
Env vars are read in [`Configure`](https://github.com/dimuon/terraform-provider-ec/blob/feature/530/migrate-to-plugin-framework/ec/provider.go#LL176C20-L176C29) function.

dimuon commented on ec/provider.go:
```diff
@@ -59,105 +70,300 @@ var (
+	var endpoint string
+	if config.Endpoint.Null {
+		endpoint = util.MultiGetenv([]string{"EC_ENDPOINT", "EC_HOST"}, api.ESSEndpoint)
+		// TODO We need to validate the endpoint here, similar to how it is done if the value is passed via terraform (isURLWithSchemeValidator)
```
Fixed.

dimuon commented on ec/provider_config.go:
```diff
@@ -18,63 +18,74 @@
-// configureAPI implements schema.ConfigureContextFunc
-func configureAPI(_ context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
-	cfg, err := newAPIConfig(d)
+func newAPIConfigLegacy(d *schema.ResourceData) (api.Config, error) {
```
I removed it.

dimuon commented on ec/internal/validators/urlvalidator.go:
```diff
@@ -0,0 +1,101 @@
+		resp.Diagnostics.AddAttributeError(
+			req.AttributePath,
+			v.Description(ctx),
+			fmt.Sprintf("URL is expected to have a valid scheme, got %v (%v)", u.Scheme, str.Value),
```
`Description` already contains the valid schemas but I'll include them here as well.

dimuon commented on ec/internal/planmodifier/default_value.go:
```diff
@@ -0,0 +1,55 @@
+func (m *defaultValueAttributePlanModifier) Modify(_ context.Context, req tfsdk.ModifyAttributePlanRequest, res *tfsdk.ModifyAttributePlanResponse) {
+	if !req.AttributeConfig.IsNull() {
+		return
+	}
```
No. I think the current logic for `include_by_default` from `ec/ecresource/trafficfilterresource/schema.go` would break if we add that condition.

dimuon replied:
Actually I tried that and the tests fail if the condition is added.

dimuon commented on ec/ecdatasource/deploymentdatasource/datasource.go:
```diff
@@ -60,92 +80,57 @@ func read(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diag
-		return err
-	}
+/*
+	TODO - see https://github.com/multani/terraform-provider-camunda/pull/16/files
```
I'm really not sure - I'm going to delete this.

dimuon commented on ec/ecresource/elasticsearchkeystoreresource/schema.go:
```diff
@@ -18,33 +18,98 @@
+				Description: "Optionally stores the remote keystore setting as a file. The default is false, which stores the keystore setting as string when value is a plain string.",
+				Optional:    true,
+				Computed:    true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{
+					planmodifier.DefaultValue(types.Bool{Value: false}),
+				},
```
@pascal-hofmann , can you please help with an answer?

pascal-hofmann replied:
From the (existing) description of this attribute: 
> Optionally stores the remote keystore setting as a file. The default is false, which stores the keystore setting as string when value is a plain string

dimuon commented on ec/ecresource/deploymentresource/apm/v1/schema.go:
```diff
@@ -0,0 +1,184 @@
+				Computed: true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{
+					planmodifier.DefaultValue(types.String{Value: "main-elasticsearch"}),
+					// resource.UseStateForUnknown(),
```
I removed it.

dimuon commented on ec/ecresource/deploymentresource/apm/v2/apm_topology.go:
```diff
@@ -0,0 +1,135 @@
+// defaultApmTopology iterates over all the templated topology elements and
+// sets the size to the default when the template size is smaller than the
+// deployment template default, the same is done on the ZoneCount.
+func defaultApmTopology(topology []*models.ApmTopologyElement) []*models.ApmTopologyElement {
```
It's a legacy from the previous code. I don't know whether it's still needed.

dimuon commented on ec/ecresource/deploymentresource/apm/v1/schema.go:
```diff
@@ -0,0 +1,184 @@
+		Optional:    true,
+		Validators:  []tfsdk.AttributeValidator{listvalidator.SizeAtMost(1)},
+		Attributes: tfsdk.ListNestedAttributes(map[string]tfsdk.Attribute{
+			// TODO
```
No - I think there is no sense to do this for v1.

dimuon replied:
No - I removed the comment.

dimuon commented:
@pascal-hofmann , can you please help with addressing PR comments to datasources and resources that were migrated by you?

pascal-hofmann commented:
> @pascal-hofmann , can you please help with addressing PR comments to datasources and resources that were migrated by you?

My laptop is overwhelmed by rendering this PR in the browser. I tried to address all comments in which I was highlighted though.

dimuon commented on ec/ecdatasource/deploymentdatasource/datasource.go:
```diff
@@ -60,92 +80,57 @@ func read(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diag
-	if err := d.Set("enterprise_search", enterpriseSearchFlattened); err != nil {
-		return err
-	}
+	diags.Append(flattenTrafficFiltering(ctx, res.Settings, &state.TrafficFilter)...)
```
@pascal-hofmann ?

pascal-hofmann replied:
Having multiple return values makes the calling code harder to read.

Regarding type safety: the type is always types.List. We also use `tfsdk.ValueFrom` inside the flattener functions, which accepts a `interface{}` too.

Feel free to change to `*types.List` or return instead.

dimuon replied:
Done

dimuon commented on ec/ecdatasource/deploymentdatasource/flatteners_apm.go:
```diff
@@ -18,84 +18,109 @@
 	"github.com/elastic/terraform-provider-ec/ec/internal/util"
 )
 
 // flattenApmResources takes in Apm resource models and returns its
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentdatasource/flatteners_elasticsearch.go:
```diff
@@ -18,147 +18,172 @@
 )
 
 // flattenElasticsearchResources takes in Elasticsearch resource models and returns its
 // flattened form.
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentdatasource/flatteners_elasticsearch.go:
```diff
@@ -18,147 +18,172 @@
-			autoscaling["max_size_resource"] = *ascale.Resource
-			autoscaling["max_size"] = util.MemoryToState(*ascale.Value)
+		var autoscaling elasticsearchAutoscalingModel
+		var empty = true
```
@pascal-hofmann ?

dimuon replied:
Done

dimuon commented on ec/ecdatasource/deploymentdatasource/flatteners_enterprise_search.go:
```diff
@@ -18,97 +18,123 @@
 )
 
 // flattenEnterpriseSearchResources takes in EnterpriseSearch resource models and returns its
 // flattened form.
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentdatasource/schema_elasticsearch.go:
```diff
@@ -18,139 +18,207 @@
-				Type:     schema.TypeBool,
-				Computed: true,
+				Type:        types.BoolType,
+				Description: "Resource kind health status.",
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentdatasource/schema_elasticsearch.go:
```diff
@@ -18,139 +18,207 @@
+			},
+			"size": {
+				Type:        types.StringType,
+				Description: "Amount of resource per topology element in the \"g\" notation.",
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentdatasource/schema_elasticsearch.go:
```diff
@@ -18,139 +18,207 @@
+			},
+			"node_type_master": {
+				Type:        types.BoolType,
+				Description: " Defines whether this node can be elected master (<7.10.0).",
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentdatasource/schema_elasticsearch.go:
```diff
@@ -18,139 +18,207 @@
+			},
+			"node_type_ingest": {
+				Type:        types.BoolType,
+				Description: "Defines whether this node can run an ingest pipeline (<7.10.0).",
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentdatasource/schema_enterprise_search.go:
```diff
@@ -18,86 +18,134 @@
-				Type:     schema.TypeBool,
-				Computed: true,
+				Type:        types.BoolType,
+				Description: "Resource kind health status.",
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentdatasource/schema_enterprise_search.go:
```diff
@@ -18,86 +18,134 @@
+			},
+			"size": {
+				Type:        types.StringType,
+				Description: "Amount of resource per topology element in the \"g\" notation.",
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentdatasource/schema_integrations_server.go:
```diff
@@ -18,72 +18,116 @@
+			},
+			"size": {
+				Type:        types.StringType,
+				Description: "Amount of resource per topology element in the \"g\" notation.",
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentdatasource/schema_kibana.go:
```diff
@@ -18,72 +18,116 @@
-				Type:     schema.TypeBool,
-				Computed: true,
+				Type:        types.BoolType,
+				Description: "Resource kind health status.",
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentdatasource/schema_kibana.go:
```diff
@@ -18,72 +18,116 @@
+			},
+			"size": {
+				Type:        types.StringType,
+				Description: "Amount of resource per topology element in the \"g\" notation.",
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentsdatasource/datasource.go:
```diff
@@ -19,108 +19,142 @@ package deploymentsdatasource
-			d.SetId(strconv.Itoa(schema.HashString(string(b))))
-		}
-	}
+/* TODO - see https://github.com/multani/terraform-provider-camunda/pull/16/files
```
@pascal-hofmann ?

dimuon replied:
I'm going to remove the TODO.

dimuon commented on ec/ecdatasource/deploymentsdatasource/schema.go:
```diff
@@ -17,154 +17,241 @@
-			Optional: true,
-			MaxItems: 1,
-			Elem:     newResourceFilters(),
+type ResourceKind int
```
@pascal-hofmann ?

dimuon replied:
As I mentioned in another comment, I moved the type to `util` package and reused it here and in the other resources.

dimuon commented on ec/ecdatasource/stackdatasource/datasource.go:
```diff
@@ -21,63 +21,113 @@ import (
+	return diags
 }
 
+/* TODO - see https://github.com/multani/terraform-provider-camunda/pull/16/files
```
@pascal-hofmann ?

dimuon replied:
I removed the TODO.

dimuon commented on ec/ecdatasource/stackdatasource/flatteners_elasticsearch.go:
```diff
@@ -18,48 +18,65 @@
-func flattenElasticsearchResources(res *models.StackVersionElasticsearchConfig) []interface{} {
-	var m = make(map[string]interface{})
+// flattenStackVersionElasticsearchConfig takes a StackVersionElasticsearchConfig and flattens it.
+func flattenStackVersionElasticsearchConfig(ctx context.Context, res *models.StackVersionElasticsearchConfig, target interface{}) diag.Diagnostics {
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/deploymentsdatasource/expanders.go:
```diff
@@ -111,42 +132,45 @@ func expandFilters(d *schema.ResourceData) (*models.SearchRequest, error) {
-			if healthy != "true" && healthy != "false" {
-				return nil, fmt.Errorf("invalid value for healthy (true|false): '%s'", healthy)
+			if filter.Healthy.Value != "true" && filter.Healthy.Value != "false" {
+				diags.AddError("invalid value for healthy", fmt.Sprintf("invalid value for healthy (true|false): '%s'", filter.Healthy.Value))
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/stackdatasource/flatteners_enterprise_search.go:
```diff
@@ -18,40 +18,55 @@
-func flattenEnterpriseSearchResources(res *models.StackVersionEnterpriseSearchConfig) []interface{} {
-	var m = make(map[string]interface{})
+// flattenStackVersionEnterpriseSearchConfig takes a StackVersionEnterpriseSearchConfig and flattens it.
+func flattenStackVersionEnterpriseSearchConfig(ctx context.Context, res *models.StackVersionEnterpriseSearchConfig, target interface{}) diag.Diagnostics {
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/stackdatasource/flatteners_kibana.go:
```diff
@@ -18,40 +18,55 @@
-func flattenKibanaResources(res *models.StackVersionKibanaConfig) []interface{} {
-	var m = make(map[string]interface{})
+// flattenStackVersionKibanaConfig takes a StackVersionKibanaConfig and flattens it.
+func flattenStackVersionKibanaConfig(ctx context.Context, res *models.StackVersionKibanaConfig, target interface{}) diag.Diagnostics {
```
@pascal-hofmann ?

dimuon commented on ec/ecdatasource/stackdatasource/schema.go:
```diff
@@ -18,103 +18,203 @@
+}
 
+func resourceKindConfigSchema(resourceKind ResourceKind) tfsdk.Attribute {
+	var names = map[ResourceKind]string{
```
@pascal-hofmann ?

dimuon replied:
I didn't manage to find appropriate type in `clilent-sdk-go` (maybe I missed it) so I moved the definition into `util` package.

dimuon commented:
> > @pascal-hofmann , can you please help with addressing PR comments to datasources and resources that were migrated by you?
> 
> My laptop is overwhelmed by rendering this PR in the browser. I tried to address all comments in which I was highlighted though.

I tagged you where it was appropriate.

dimuon commented on ec/ecresource/deploymentresource/kibana/v1/schema.go:
```diff
@@ -0,0 +1,160 @@
+			},
+			"config": {
+				Optional: true,
+				// TODO
```
I think it doesn't make sense for v1. I'll remove the comment.

dimuon commented on ec/ecresource/deploymentresource/elasticsearch/v2/elasticsearch_topology.go:
```diff
@@ -0,0 +1,394 @@
+	Autoscaling             *ElasticsearchTopologyAutoscaling `tfsdk:"autoscaling"`
+}
+
+func CreateTierForTest(tierId string, tier ElasticsearchTopology) *ElasticsearchTopology {
```
It's called from other packages tests. I cannot come up with a better place for the func unfortunately.

dimuon replied:
I moved it to a test file.

dimuon commented on ec/ecresource/deploymentresource/utils/node_types_to_node_roles.go:
```diff
@@ -0,0 +1,94 @@
+		return false, diags
+	}
+
+	return useNodeRoles && convertLegacy, nil
```
I brought the old logic back. @tobio , please have a look.

dimuon commented on ec/ecresource/deploymentresource/utils/node_types_to_node_roles.go:
```diff
@@ -0,0 +1,94 @@
+	DataTiersVersion = semver.MustParse("7.10.0")
+)
+
+func UseNodeRoles(stateVersion, planVersion types.String) (bool, diag.Diagnostics) {
```
I added tests and moved the majority of files from the folder to other packages.

dimuon commented on ec/ecresource/deploymentresource/elasticsearch/v2/node_types_plan_modifier.go:
```diff
@@ -0,0 +1,124 @@
+	return "Use current state if it's still valid."
+}
+
+func isAttributeChanged(ctx context.Context, p path.Path, req tfsdk.ModifyAttributePlanRequest) (bool, diag.Diagnostics) {
```
Done.

dimuon commented on ec/ecresource/deploymentresource/elasticsearch/v2/elasticsearch_topology.go:
```diff
@@ -0,0 +1,394 @@
+		topology.NodeTypeIngest.Value != "" ||
+		topology.NodeTypeMaster.Value != "" ||
+		topology.NodeTypeMl.Value != ""
+}
```
It's used now once I brought back the old logic for migrating from `node_types` to `node_roles`.

dimuon commented on ec/ecresource/deploymentresource/elasticsearch/v2/elasticsearch_topology.go:
```diff
@@ -0,0 +1,394 @@
+	}
+
+	return topology, nil
+}
```
It's used now.

dimuon commented on ec/ecdatasource/deploymentdatasource/flatteners_elasticsearch.go:
```diff
@@ -18,147 +18,172 @@
-			m["node_roles"] = schema.NewSet(schema.HashString, util.StringToItems(
-				topology.NodeRoles...,
-			))
+			diags.Append(tfsdk.ValueFrom(ctx, util.StringToItems(topology.NodeRoles...), types.SetType{ElemType: types.StringType}, &model.NodeRoles)...)
```
I removed `util.StringToItems` func

dimuon commented on ec/ecresource/deploymentresource/deployment/v2/deployment.go:
```diff
@@ -0,0 +1,566 @@
+		return nil, diagsnostics
+	}
+
+	useNodeRoles, err := utils.CompatibleWithNodeRoles(version)
```
I reverted back the previous logic.

dimuon replied:
@tobio, please have a look.

dimuon commented:
@pascal-hofmann , can you sign the CLA? If not I will have to proceed with the git history rewriting - I'll modify your commits to use the email address that you used to sign the CLA. Still, I prefer not to do that.

dimuon commented:
@tobio , I believe all your comments were addressed - either with a fix or with a reasoning behind the made change. Can you please have a look at the PR again?

tobio commented on ec/ecresource/deploymentresource/kibana/v2/schema.go:
```diff
@@ -0,0 +1,142 @@
+					},
+					"user_settings_yaml": {
+						Type:        types.StringType,
+						Description: `An arbitrary YAML object allowing ECE admins owners to set clusters' parameters (only one of this and 'user_settings_override_json' is allowed), ie in addition to the documented 'system_settings'. (This field together with 'system_settings' and 'user_settings*' defines the total set of resource settings)`,
```
`user_settings_yaml` <> `user_settings_override_yaml`

tobio commented on ec/ecresource/deploymentresource/read.go:
```diff
@@ -92,3 +173,53 @@ func deploymentNotFound(err error) bool {
+	return nil
+}
+
+func HasRunningResources(res *models.DeploymentGetResponse) bool {
```
```suggestion
func hasRunningResources(res *models.DeploymentGetResponse) bool {
```

dimuon replied:
We cannot rename it - it's tested in `deploymentresource_test`. And the test package is named differently because it imports `github.com/elastic/terraform-provider-ec/ec/ecresource/deploymentresource`.

dimuon replied:
So it would be a circular dependency if the test package was named `deploymentresource`.

tobio commented on ec/internal/converters/extract_endpoint.go:
```diff
@@ -0,0 +1,62 @@
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+// FlattenClusterEndpoint receives a ClusterMetadataInfo, parses the http and
```
This comment is out of date

tobio commented on ec/internal/converters/extract_endpoint.go:
```diff
@@ -0,0 +1,62 @@
+	return
+}
+
+// FlattenClusterEndpoint receives a ClusterMetadataInfo, parses the http and
```
Similarly out of date

tobio reviewed (commented):
Bunch of minor comments, otherwise looking good. I'll try and do some more specific testing tomorrow.

dimuon commented on ec/ecresource/deploymentresource/elasticsearch/v2/schema.go:
```diff
@@ -0,0 +1,471 @@
+	}
+}
+
+func ElasticsearchRemoteClusterSchema() tfsdk.Attribute {
```
It's used in `ec/ecresource/deploymentresource/deployment/v2/elasticsearch_remote_cluster_payload_test.go`.

dimuon commented on ec/ecresource/deploymentresource/elasticsearch/v2/topology_plan_modifier.go:
```diff
@@ -0,0 +1,97 @@
+)
+
+// Use current state for a topology's attribute if the topology's state is not nil and the template attribute has not changed
+func UseTopologyStateForUnknown(topologyAttributeName string) tfsdk.AttributePlanModifier {
```
it's used in `ec/ecresource/deploymentresource/elasticsearch/v2/topology_plan_modifier_test.go` that is `v2_test` package.

dimuon commented on ec/ecresource/deploymentresource/elasticsearch/v2/node_roles.go:
```diff
@@ -0,0 +1,203 @@
+	}
+
+	// if template changed return
+	templateChanged, diags := attributeChanged(ctx, path.Root("deployment_template_id"), req)
+
+	resp.Diagnostics.Append(diags...)
+
+	if diags.HasError() {
+		return false, false
+	}
+
+	if templateChanged {
+		return false, false
+	}
```
The function is used only in `node_roles` and `node_types` plan modifies. The `if` case prevents the plan modifies from using the current state as a planned value. Without that, once the template is modified, the plan modifier would still use the current state. I think it's more about precaution - once the template is changed, the plan output suggests that future values for `node_roles` and `node_types` are "known after apply". They still can be the same as the current state values though. The problem with plan modifies is that if they make wrong decision, it becomes an error in the provider while if they doesn't suggest any value (effectively keeping "known after apply"), it could be just extra output in the plan.

dimuon commented on ec/ecresource/deploymentresource/elasticsearch/v2/node_roles.go:
```diff
@@ -0,0 +1,203 @@
+	"context"
+	"fmt"
+
+	"github.com/blang/semver"
```
I'll do this.

dimuon replied:
https://github.com/elastic/terraform-provider-ec/issues/577

dimuon commented:
@tobio , just FYI - I've updated the PR description where mentioned the opened defect to TF core where we also discovered that sets cannot be used with computed and optional fields that prevent us from using sets for Elasticsearch topology.

tobio commented on ec/internal/converters/extract_endpoint.go:
```diff
@@ -25,7 +25,7 @@ import (
 
-// FlattenClusterEndpoint receives a ClusterMetadataInfo, parses the http and
+// ExtractEndpointsToTypes receives a ClusterMetadataInfo, parses the http and
 // https endpoints and returns a map with two keys: `http_endpoint` and
```
These functions no longer return a map either, wasn't just the name.

dimuon replied:
I removed the comments. The function signatures are self descriptive.

tobio commented on ec/ecresource/deploymentresource/apm/v1/schema.go:
```diff
@@ -0,0 +1,159 @@
+				Optional:    true,
+			},
+			"user_settings_yaml": {
+				Type:        types.StringType,
```
user_settings_yaml <> user_settings_override_yaml

tobio commented on ec/ecresource/deploymentresource/apm/v2/schema.go:
```diff
@@ -0,0 +1,143 @@
+			},
+			"user_settings_yaml": {
+				Type:        types.StringType,
+				Description: `An arbitrary YAML object allowing ECE admins owners to set clusters' parameters (only one of this and 'user_settings_override_json' is allowed), ie in addition to the documented 'system_settings'. (This field together with 'system_settings' and 'user_settings*' defines the total set of resource settings)`,
```
user_settings_yaml <> user_settings_override_yaml

dimuon replied:
Oh, sorry - I overlooked this.

pascal-hofmann commented:
As far as I know the CLA has been signed. I guess my comment will re-trigger the CLA check?

tobio commented:
@pascal-hofmann I can see you've signed the CLA, however it looks like these commits are using a different email address (`pascal.hofmann.ext@gec.io`), not tied to your GH account. 

https://github.com/elastic/terraform-provider-ec/pull/567/commits/2194ad39245fdc4fc9e57743f2aab3ef626f7ab4
https://github.com/elastic/terraform-provider-ec/pull/567/commits/8a7997048d18307bd64a6f8058c707d4ba75b20b
https://github.com/elastic/terraform-provider-ec/pull/567/commits/c197f28cdf5834ad1cd670ff8f3076753afd05dc
https://github.com/elastic/terraform-provider-ec/pull/567/commits/b00e28a8228e2c40006915bec770c3bc3d5fc4a3

@dimuon I guess you could reset the author on the above commits to use `mail@pascalhofmann.de` in order to get past the check. FWIW manually triggering the CLA checker on this PR timed out, maybe there's a longer timeout on the automatic status check or we just force merge it 🤷.

pascal-hofmann commented:
I've added the address to my github account. I don't think rewriting the commits is the right thing to do.

dimuon commented:
I've just rerun the checker. It looks good know 👍

dimuon commented:
I still would like to update the TF Framework to the latest version (v.1.1.1) before the merge. I've started working on this but it'll take some time - there are a number of breaking changes between 0.14.0 and 1.1.1.

tobio commented:
https://www.youtube.com/watch?v=fR4HjTH_fTM

dimuon commented:
@tobio , I've updated the framework to v1.1.1. Would you mind to have a look at the [commit](https://github.com/elastic/terraform-provider-ec/pull/567/commits/06bf06e4431087a53a0d94971f7391dbef6d20ae)?

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/node_roles_plan_modifier.go:
```diff
@@ -20,17 +20,19 @@ package v2
 )
 
-func UseNodeRolesDefault() tfsdk.AttributePlanModifier {
+func UseNodeRolesDefault() nodeRolesDefault {
```
The previous return value was also effectively a type assertion. 

Might be worth adding something like this to the top of this file:

```
var _ planmodifier.Set = nodeRolesDefault{}
```

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/node_types_plan_modifier.go:
```diff
@@ -20,17 +20,19 @@ package v2
 )
 
-func UseNodeTypesDefault() tfsdk.AttributePlanModifier {
+func UseNodeTypesDefault() nodeTypesDefault {
```
The previous return value was also effectively a type assertion. 

Might be worth adding something like this to the top of this file:

```
var _ planmodifier.String = nodeTypesDefault{}
```

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/topology_plan_modifier.go:
```diff
@@ -23,59 +23,77 @@ import (
 
 // Use current state for a topology's attribute if the topology's state is not nil and the template attribute has not changed
-func UseTopologyStateForUnknown(topologyAttributeName string) tfsdk.AttributePlanModifier {
+func UseTopologyStateForUnknown(topologyAttributeName string) useTopologyState {
```
Similarly, might be worth adding some type assertions for this struct at the top

tobio commented on ec/ecresource/deploymentresource/elasticsearch/v2/topology_plan_modifier.go:
```diff
@@ -23,59 +23,77 @@ import (
-	// if the config is the unknown value, use the unknown value otherwise, interpolation gets messed up
-	if req.AttributeConfig.IsUnknown() {
-		return
+type PlanModifierResponse interface {
```
This appears to be unused?

dimuon replied:
Good catch 👍

tobio commented on ec/ecresource/deploymentresource/observability/v1/schema.go:
```diff
@@ -18,47 +18,31 @@
-				PlanModifiers: []tfsdk.AttributePlanModifier{
-					planmodifier.DefaultValue(types.Bool{Value: true}),
+				"logs": schema.BoolAttribute{
+					Optional: true,
```
Why do we drop the default values here but not in the v2 schema?

dimuon replied:
I think it doesn't make sense to have plan modifiers and validators in v1 - the only need of v1 is to support state upgrade (once we decide to implement it) so currently v1 isn't used at all. Also _AFAIK_ both plan modifiers and validators aren't invoked during state upgrade.

tobio commented on ec/internal/planmodifiers/bool_default_value.go:
```diff
@@ -0,0 +1,57 @@
+	value bool
+}
+
+func BoolDefaultValue(v bool) planmodifier.Bool {
```
We should be consistent with the ES plan modifiers (i.e return a concrete type), also probably should have the `planmodifier.Bool` type assertion at the top of this file.

tobio commented on ec/internal/planmodifiers/string_default_value.go:
```diff
@@ -0,0 +1,57 @@
+	value string
+}
+
+func StringDefaultValue(v string) planmodifier.String {
```
Similarly, should return a concrete type and include a type assertion above.

tobio commented on ec/internal/validators/knownvalidator.go:
```diff
@@ -53,6 +50,6 @@ func (v knownValidator) Validate(ctx context.Context, req tfsdk.ValidateAttribut
 //
 // Null (unconfigured) values are skipped.
-func Known() tfsdk.AttributeValidator {
+func Known() validator.String {
```
Similarly should return a concrete type and include a type assertion at the top of this file.

tobio commented on ec/internal/validators/urlvalidator.go:
```diff
@@ -23,79 +23,65 @@ import (
 }
 
-func IsURLWithSchemeValidator(validSchemes []string) tfsdk.AttributeValidator {
+func IsURLWithSchemeValidator(validSchemes []string) validator.String {
```
Similarly should return a concrete type and include a type assertion at the top of this file.

tobio reviewed (commented):
LGTM. Some minor nits. Some cases where we should be following 'accept interfaces, return structs'

dimuon commented on ec/ecdatasource/deploymentsdatasource/schema.go:
```diff
@@ -22,166 +22,146 @@ import (
-				PlanModifiers: []tfsdk.AttributePlanModifier{
-					planmodifier.DefaultValue(types.Int64{Value: 100}),
-				},
+				// PlanModifiers: []tfsdk.AttributePlanModifier{
```
I killed it. Plan modifiers aren't supported by datasources.

dimuon commented on ec/ecdatasource/deploymentdatasource/flatteners_elasticsearch.go:
```diff
@@ -162,25 +158,23 @@ func flattenElasticsearchTopology(ctx context.Context, plan *models.Elasticsearc
 
 		if hasAutoscalingModel {
-			diags.Append(tfsdk.ValueFrom(ctx, []elasticsearchAutoscalingModel{autoscaling}, elasticsearchAutoscalingListType(), &model.Autoscaling)...)
+			var d diag.Diagnostics
+			model.Autoscaling, d = types.ListValueFrom(ctx, elasticsearchAutoscalingElemType(), []elasticsearchAutoscalingModel{autoscaling})
+			diags.Append(d...)
```
We cannot do like that - `model.Autoscaling` cannot be redeclared.

tobio replied:
🤦

dimuon commented:
I’ve updated the TF framework version till the latest one (v1.1.1) and came across a [defect](https://github.com/hashicorp/terraform-plugin-framework/issues/644) in the Framework that prevents us from releasing based on v1.1.1 (see our acceptance test [failures](https://devops-ci.elastic.co/job/elastic+terraform-provider-ec+pull-request/869/console) - all errors with `elasticsearch.trust_account`'s unexpected new value are caused by the defect). The defect is already fixed in the Framework main branch and should be in [v1.2.0](https://github.com/hashicorp/terraform-plugin-framework/milestone/26).

I plan to revert the latest commits that updates the Framework version and release the provider based on Framework v0.14.0 that was used during the development and was tested the most.

We will update the TF Framework till v1.2.0 in the subsequent releases.

joelsdc commented:
This merge breaks datasources. Please undo or fix ASAP.

dimuon commented:
> This merge breaks datasources. Please undo or fix ASAP.

@joelsdc , can you please provide more details? What is the error message? Also please share your configuration.

dimuon commented:
@joelsdc , can you please also share the version of the Terraform CLI.

joelsdc commented:
Hi @dimuon, I've opened issue #595 

I'm using:
```
Terraform v1.3.9
on darwin_arm64
```
